pkg/operator/ceph/file/spec.go

```diff
@@ -0,0 +1,166 @@
+	opspec "github.com/rook/rook/pkg/operator/ceph/spec"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	mdsDaemonCommand = "ceph-mds"
+)
+
+func makeDeployment(
+	clientset kubernetes.Interface,
+	fs cephv1beta1.Filesystem,
+	filesystemID,
+	version string,
+	hostNetwork bool,
+	ownerRefs []metav1.OwnerReference,
+) *extensions.Deployment {
+	deployment := &extensions.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instanceName(fs),
+			Namespace: fs.Namespace,
+		},
+	}
+	k8sutil.SetOwnerRefs(clientset, fs.Namespace, &deployment.ObjectMeta, ownerRefs)
+
+	podSpec := v1.PodSpec{
+		InitContainers: []v1.Container{
+			makeConfigInitContainer(fs, filesystemID, version),
+		},
+		Containers: []v1.Container{
+			makeMdsDaemonContainer(fs, filesystemID, version),
+		},
+		RestartPolicy: v1.RestartPolicyAlways,
+		Volumes:       opspec.PodVolumes(""),
+		HostNetwork:   hostNetwork,
+	}
+	if hostNetwork {
+		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+	fs.Spec.MetadataServer.Placement.ApplyToPodSpec(&podSpec)
+
+	podTemplateSpec := v1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        instanceName(fs),
+			Labels:      getLabels(fs),
+			Annotations: map[string]string{},
+		},
+		Spec: podSpec,
+	}
+
+	// double the number of MDS instances for failover
+	mdsCount := fs.Spec.MetadataServer.ActiveCount * 2
+
+	deployment.Spec = extensions.DeploymentSpec{Template: podTemplateSpec, Replicas: &mdsCount}
+
+	return deployment
+}
+
+func makeConfigInitContainer(fs cephv1beta1.Filesystem, filesystemID, version string) v1.Container {
+	return v1.Container{
+		Name: opspec.ConfigInitContainerName,
+		Args: []string{
+			"ceph",
+			mdsdaemon.InitCommand,
+			fmt.Sprintf("--config-dir=%s", k8sutil.DataDir),
+			fmt.Sprintf("--filesystem-id=%s", filesystemID),
+			fmt.Sprintf("--active-standby=%s", strconv.FormatBool(fs.Spec.MetadataServer.ActiveStandby)),
+		},
+		Image: k8sutil.MakeRookImage(version),
+		Env: []v1.EnvVar{
+			{Name: "ROOK_POD_NAME", ValueFrom: &v1.EnvVarSource{FieldRef: &v1.ObjectFieldSelector{FieldPath: "metadata.name"}}},
+			k8sutil.PodIPEnvVar(k8sutil.PrivateIPEnvVar),
+			k8sutil.PodIPEnvVar(k8sutil.PublicIPEnvVar),
+			opmon.ClusterNameEnvVar(fs.Namespace),
+			opmon.EndpointEnvVar(),
+			opmon.SecretEnvVar(),
+			opmon.AdminSecretEnvVar(),
+			k8sutil.ConfigOverrideEnvVar(),
+		},
+		VolumeMounts: opspec.RookVolumeMounts(),
+		Resources:    fs.Spec.MetadataServer.Resources,
+	}
+}
+
+func makeMdsDaemonContainer(fs cephv1beta1.Filesystem, filesystemID, version string) v1.Container {
+	return v1.Container{
+		Name: "mgr",
+		Command: []string{
+			mdsDaemonCommand,
+		},
+		Args: []string{
+			"--foreground",
+			"--id",
+			// TODO: We can't wait for the POD id here;
+			//       we need to do something different to have the name beforehand
```

Comment:
@travisn, I've hit a situation with the MDS, and I'd like to get your advice on how best to proceed:
My understanding of the MDS is that currently, it uses its own pod name to figure out how to run the process. This doesn't work for running the daemon directly from the spec (above) since the name isn't yet known. Should I rework this to follow the same pattern as the mons/mgrs to create a deployment for each mds pod? And I might additionally be able to combine the main MDS pod and it's active standby pod into one deployment if the arguments to both of them are the same. Or are there  things I can do with the deployment/replicaset to still have this info beforehand?

@liewegas might also have good insights.

Reply:
If the `-id` command line arg is not available as a config setting, then the only option I can think of is to have the operator create one deployment per instance of MDS as you suggested. I do remember a conversation with @liewegas that we wanted this behavior to give us consistent naming of the mds pods with the a, b, c suffix. For the upgrade scenario, we would just need to delete the old mds deployment if it exists, then proceed creating the new one. This pattern is already in place for deleting a legacy mon replicaset. 

@liewegas as I recall, each time a new mds pod is started, a new mds id should be used. Is that true or am I remembering something else? In that case, we would have an issue if k8s fails over the pod to another node and starts it with the same id.